### PR TITLE
OKTA-530556 : Password validation against username implementation

### PR DESCRIPTION
**Pull request by glenfannin-okta**
## Description:
Purpose of this is to validate the new password against parts of the username.

## PR Checklist

- [x] Have you verified the basic functionality for this change?
- [x] Did you add tests, as appropriate, following our [Automated Test guidelines](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/2676497890/Automated+Testing+in+the+Signin+Widget)?
- [x] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?
- [ ] Did you verify the change by running [downstream monolith artifact](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/102897979/Sign-in+Widget+Development#Sign-inWidgetDevelopment-Instructionstocreateandrunthedownstreamartifact(d16t))? (Provide link to build below)
- [x] Does this PR include noticeable changes to the UI? (If yes, attach screenshots/video below)

### Issue:

- [OKTA-530556](https://oktainc.atlassian.net/browse/OKTA-530556)

### Screenshot/Video:
Full username:
image

Partial:
image

image

### CODE REVIEW

File: src/v3/src/components/Widget/index.tsx [hunk 1]
```diff
@@ -286,7 +286,8 @@ export const Widget: FunctionComponent<WidgetProps> = (widgetProps) => {
     >
       {/* Note that we need two theme providers until we fully migrate to odyssey-mui */}
       <MuiThemeProvider theme={mapMuiThemeFromBrand(brandColors)}>
-        <ScopedCssBaseline>
+        {/* the style is to allow the widget to inherit the parent's bg color */}
+        <ScopedCssBaseline sx={{ backgroundColor: 'inherit' }}>
```
Comment (shawyu-okta) [inline src/v3/src/components/Widget/index.tsx:290]: Is this related to this ticket?

File: src/v3/src/util/passwordUtils.test.ts [hunk 2]
```diff
@@ -205,9 +205,9 @@ describe('PasswordUtils Tests', () => {
     userInfo = { identifier: 'tester.user+123@okta.com' };
 
     expect(validatePassword('tester.user', userInfo, { complexity: { excludeUsername: true } })?.excludeUsername)
-      .toEqual(true);
+      .toEqual(false);
     expect(validatePassword('tester.user@okta.com', userInfo, { complexity: { excludeUsername: true } })?.excludeUsername)
-      .toEqual(true);
+      .toEqual(false);
```
Comment (shawyu-okta) [inline src/v3/src/util/passwordUtils.test.ts:210]: 👌

**Review by shawyu-okta — APPROVED** [review]

File: src/v3/src/components/Widget/index.tsx (hunk 1 above)
Comment (glenfannin-okta) [reply src/v3/src/components/Widget/index.tsx:290]: Oops, no, I was working with Lee on something and forgot to remove this 😅 will remove (this should be added in his branch for the ticket he is working on).

File: src/v3/src/util/passwordUtils.ts [hunk 3]
```diff
@@ -55,6 +55,37 @@ const minSymbolValidator = (password: string, limit: unknown): boolean => (
...
 
+/*
+* Copied from backend logic
+* See: okta-core - PasswordUtil.java#L72-L97
+*/
+const getParts = (attributeVal: string): string[] => {
+  const MIN_PARTS_LENGTH = 4;
+  const parts: string[] = [];
+  const delimiters = new Set<string>([',', '.', '-', '_', '#', '@']);
+  const characters = attributeVal.split('');
```
[the remark this answers is not in the document]
Comment (glenfannin-okta) [reply src/v3/src/util/passwordUtils.ts:66]: I could just use `for ( const char of val) { ...` however, when I do I get the below lint error: 

```
iterators/generators require regenerator-runtime, which is too heavyweight for this guide to allow them. Separately, loops should be avoided in favor of array iterations.eslint[no-restricted-syntax](https://eslint.org/docs/rules/no-restricted-syntax)
```

But `Array.from` works just as well and supports multi-byte characters

**Review by shawyu-okta — APPROVED** [review]

File: src/v3/src/components/Widget/index.tsx [hunk 4]
```diff
@@ -286,8 +286,7 @@ export const Widget: FunctionComponent<WidgetProps> = (widgetProps) => {
     >
       {/* Note that we need two theme providers until we fully migrate to odyssey-mui */}
       <MuiThemeProvider theme={mapMuiThemeFromBrand(brandColors)}>
-        {/* the style is to allow the widget to inherit the parent's bg color */}
-        <ScopedCssBaseline sx={{ backgroundColor: 'inherit' }}>
```
Comment (shawyu-okta) [inline src/v3/src/components/Widget/index.tsx:290]: looks like there was some conflict that resolved the wrong way after this was merged in the other PR

Comment (glenfannin-okta) [reply src/v3/src/components/Widget/index.tsx:290]: Arrgh, git thinks its smarter than me 😅 

TY!
